File: code/python/ch07_counts_poisson.py
"""
Chapter 7 - Statistical foundations
Crime counts, overdispersion, and a Poisson regression. Self-contained.
"""
# --- book code (Chapter 7) -------------------------------------------------
# Python: check overdispersion, then fit a Poisson model
import numpy as np, pandas as pd
import statsmodels.api as sm, statsmodels.formula.api as smf
rng = np.random.default_rng(7)
area = pd.DataFrame({"deprivation": rng.normal(0, 1, 500)})
mu = np.exp(1.0 + 0.5 * area["deprivation"])     # true relationship
area["crimes"] = rng.poisson(mu)
print("variance / mean =", round(area.crimes.var()/area.crimes.mean(), 2))
m = smf.glm("crimes ~ deprivation", data=area,
            family=sm.families.Poisson()).fit()
print(m.params.round(3))

File: code/r/ch07_counts_poisson.R
# Chapter 7 - Statistical foundations
# Crime counts, overdispersion, and a Poisson regression.

# setup: build the same synthetic area frame as the Python version
set.seed(7)
area <- data.frame(deprivation = rnorm(500))
mu <- exp(1.0 + 0.5 * area$deprivation)
area$crimes <- rpois(500, mu)
cat("variance / mean =", round(var(area$crimes) / mean(area$crimes), 2), "\n")

# --- book code (Chapter 7) -------------------------------------------------
# R: the same, moving to negative binomial if dispersion is high
m <- glm(crimes ~ deprivation, data = area, family = poisson)
summary(m)                       # compare residual deviance to its d.f.
# MASS::glm.nb(crimes ~ deprivation, data = area)  # if overdispersed
